### PR TITLE
Cleanup CachedContainer invalidation

### PR DESCRIPTION
**Pull request by staabm**
- centralize cache invalidation

### CODE REVIEW

File: src/Kernel/CacheInvalidatingContainer.php [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    }
+
+    public function get(string $id, int $invalidBehavior = self::EXCEPTION_ON_INVALID_REFERENCE): ?object
+    {
+        try {
+            return $this->wrapped->get($id, $invalidBehavior);
+        } catch (\Throwable $e) {
+            RectorKernel::clearCache();
+
+            throw $e;
```
Comment (staabm) [inline src/Kernel/CacheInvalidatingContainer.php]: I though about automatically creating a fresh uncached container in this case, but it did not yet work for me.

will look into it again in another PR

File: src/Kernel/CacheInvalidatingContainer.php [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    public function get(string $id, int $invalidBehavior = self::EXCEPTION_ON_INVALID_REFERENCE): ?object
+    {
+        try {
+            return $this->wrapped->get($id, $invalidBehavior);
+        } catch (\Throwable $throwable) {
+            // clear compiled container cache, to trigger re-discovery
+            RectorKernel::clearCache();
+
+            throw new \RuntimeException(
+                'Container cache is outdated and was cleared. please re-run the command.',
```
Comment (staabm) [inline src/Kernel/CacheInvalidatingContainer.php:30]: lets throw with a more explicit error message, so people know about the "auto-healing" capabilities

**Review by samsonasik — CHANGES_REQUESTED** [review]
can "v8" container cache be named generic name or still need increase the value?

**Comment by staabm** [comment]
> can "v8" container cache be named generic name or still need increase the value?

[I guess you are talking about the cache-key](https://github.com/rectorphp/rector-src/blob/c0a042c3b68eb657d7e525ae2c2648d0ec96fb19/src/Kernel/RectorKernel.php#L20). using a generic string for this cache-key does not make sense.

I think there is still [room for improvement](https://github.com/rectorphp/rector-src/pull/3867#discussion_r1194813798) but as of now, it should work as far as I know.

we need to try it for a bit longer to get an idea which cases are still problematic and which force us to increase the kernel-cache-key. if we have concrete repro steps we can think about how to progress further

**Review by samsonasik — APPROVED** [review]
Ok, lets merge it so we have faster feedback 👍

**Comment by samsonasik** [comment]
@staabm I need to revert this as I tried in CodeIgniter4 project with clear-cache and got error:

```
Fatal error: Uncaught Error: Class "RectorPrefix202305\Symplify\SmartFileSystem\SmartFileSystem" not found in /Users/samsonasik/www/CodeIgniter4/vendor/rector/rector/src/Kernel/CachedContainerBuilder.php:71
Stack trace:
#0 /Users/samsonasik/www/CodeIgniter4/vendor/rector/rector/src/Kernel/RectorKernel.php(76): Rector\Core\Kernel\CachedContainerBuilder->clearCache()
#1 /Users/samsonasik/www/CodeIgniter4/vendor/rector/rector/src/Console/Command/ProcessCommand.php(144): Rector\Core\Kernel\RectorKernel::clearCache()
```

**Pull request merged**